scipy/optimize/_qap.py

```diff
@@ -0,0 +1,437 @@
+    elif seed.shape[1] != 2:
+        msg = "Seed array entry must have two columns"
+    elif not (seed >= 0).all():
+        msg = "Seed array entries must be greater than or equal to zero"
+    elif not (seed <= (cost_matrix.shape[0] - 1)).all():
+        msg = "Seed array entries must be less than the number of nodes"
+    elif not len(set(seed[:, 0])) == len(seed[:, 0]) or not \
+            len(set(seed[:, 1])) == len(seed[:, 1]):
+        msg = "Seed column entries must be unique"
+    elif init not in {'barycenter', 'rand'}:
+        msg = "Invalid 'init_method' parameter string"
+    if msg is not None:
+        raise ValueError(msg)
+
+    # TypeError check
+    if type(n_init) is not int or n_init <= 0:
+        msg = "'n_init' must be a positive integer"
+    elif maxiter <= 0 or type(maxiter) is not int:
+        msg = "'max_iter' must be a positive integer"
+    elif type(shuffle_input) is not bool:
+        msg = "'shuffle_input' must be a boolean"
+    elif eps <= 0 or type(eps) is not float:
+        msg = "'eps' must be a positive float"
+    elif type(maximize) is not bool:
+        msg = "'maximize' must be a boolean"
+    if msg is not None:
+        raise TypeError(msg)
+
+    n = cost_matrix.shape[0]  # number of vertices in graphs
+    n_seeds = seed.shape[0]  # number of seeds
+    n_unseed = n - n_seeds
+
+    perm_inds = np.zeros(n)
+
+    obj_func_scalar = 1
+    if maximize:
+        obj_func_scalar = -1
+    score = obj_func_scalar * math.inf
+
+    seed_dist_c = np.setdiff1d(range(n), seed[:, 1])
+    if shuffle_input:
+        seed_dist_c = np.random.permutation(seed_dist_c)
+        # shuffle_input to avoid results from inputs that were already matched
+
+    seed_cost_c = np.setdiff1d(range(n), seed[:, 0])
+    permutation_cost = np.concatenate([seed[:, 0],
+                                       seed_cost_c], axis=None).astype(int)
+    permutation_dist = np.concatenate([seed[:, 1],
+                                       seed_dist_c], axis=None).astype(int)
+    cost_matrix = cost_matrix[np.ix_(permutation_cost, permutation_cost)]
+    dist_matrix = dist_matrix[np.ix_(permutation_dist, permutation_dist)]
+
+    # definitions according to Seeded Graph Matching [2].
+    A11 = cost_matrix[:n_seeds, :n_seeds]
+    A12 = cost_matrix[:n_seeds, n_seeds:]
+    A21 = cost_matrix[n_seeds:, :n_seeds]
+    A22 = cost_matrix[n_seeds:, n_seeds:]
+    B11 = dist_matrix[:n_seeds, :n_seeds]
+    B12 = dist_matrix[:n_seeds, n_seeds:]
+    B21 = dist_matrix[n_seeds:, :n_seeds]
+    B22 = dist_matrix[n_seeds:, n_seeds:]
+
+    for i in range(n_init):
+        # setting initialization matrix
+        if init == "rand":
+            K = np.random.rand(n_unseed, n_unseed)
+            # generate a nxn matrix where each entry is a random integer [0, 1]
+            for i in range(10):  # perform 10 iterations of Sinkhorn balancing
+                K = _doubly_stochastic(K)
```

Comment:
_doubly_stochastic already does iteration internally so the loop will not do anything extra improvement.